pkg/kn/commands/plugin/path_verifier.go

```diff
@@ -0,0 +1,102 @@
+// Copyright © 2018 The Knative Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+// PathVerifier receives a path and determines if it is valid or not
+type PathVerifier interface {
+	// Verify determines if a given path is valid
+	Verify(path string) []error
+}
+
+// CommandOverrideVerifier verifies that existing kn commands are not overriden
+type CommandOverrideVerifier struct {
+	Root        *cobra.Command
+	SeenPlugins map[string]string
+}
+
+// Verify implements PathVerifier and determines if a given path
+// is valid depending on whether or not it overwrites an existing
+// kn command path, or a previously seen plugin.
+func (v *CommandOverrideVerifier) Verify(path string) []error {
+	if v.Root == nil {
+		return []error{fmt.Errorf("unable to verify path with nil root")}
+	}
+
+	// extract the plugin binary name
+	segs := strings.Split(path, "/")
+	binName := segs[len(segs)-1]
+
+	cmdPath := strings.Split(binName, "-")
+	if len(cmdPath) > 1 {
+		// the first argument is always "kn" for a plugin binary
+		cmdPath = cmdPath[1:]
+	}
+
+	errors := []error{}
+	isExec, err := isExecutable(path)
```

Comment:
That’s fair, though, the point of listing the plugins is to also give some “status” for each since if listed user should be able to run it. The question then becomes do we do more or less checks for user. I am always in favor of helping the user since a non-executable plugin for the current user will not work. So listing it w/o warning or error is misleading at best.

Reply:
Ok, then please make the `isExecutable()` complete then and verify that the current user can actually really execute it and not only checking if _somebody_ can execute it.

I think the check that the go library does for is good enough (and complete).

Reply:
This goes to something I don't think I've said very clearly. There's a contract for what a plug-in is: 1) starts with `kn-`, 2) can be found via PATH or well-known dir, 3) is marked as executable.

All 3 conditions much be met. However, how you view these requires is very important. For example, I look at them as "the user must meet all 3 otherwise kn will ignore the file". I believe @maximilien is looking it as "any file that meets 1) and 2) must also meet 3) otherwise it's an error".  IOW, 1) and 2) are triggers for kn to believe that the file is meant to it and therefore the user made a mistake by not doing 3).  While to me, if the user chooses not to do 3) it's because they wanted to. Meaning, they purposely chose not to full-fill the contract because they don't want it to be considered.

Is the contract 1+2 with 3 being a check, or is the contract 1+2+3 ?   IMO, it's the latter.

Reply:
As mentioned I wouldn't make it that complicated and just let the error pop up when the plugin is executed. So just don't check for it. Its a rare case that a file in a directory within the $PATH is not executable. (except maybe when the user directly downloaded into the directory via curl)

But this here is not mission-critical, happy to accept it as is (when we got the other issues sorted out). We can refine later.

Reply:
See now.